### PR TITLE
Unit tests for HandleHolderBase

### PR DESCRIPTION
**Pull request by alxrsngrtn**

### CODE REVIEW

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  @Test
+  fun dispatcher_someHandles_getsFirstHandlesDispatcher() {
+    val handleHolder = DummyHandleHolder()
+    val firstHandle = mock<Handle> {
+      on { dispatcher } doReturn DUMMY_DISPATCHER_1
+    }
+    val secondHandle = mock<Handle> {
+      on { dispatcher } doReturn DUMMY_DISPATCHER_2
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:77]: This one is interesting... I think technically it should be an error to try and configure a HandleHolder this way. All handles should have the same dispatcher. I wonder if we should consider adding something that verifies that as handles are set... (not in this PR of course)

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@OptIn(ExperimentalCoroutinesApi::class)
+@RunWith(JUnit4::class)
+class HandleHolderBaseTest {
+
+  @Test
+  fun handles_accessMissingHandle_throwsNoElemException() {
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:32]: I'd argue that this test isn't needed. You're testing a structure that's created solely for the purpose of allowing this test to pass.

The GeneratedHandleHolder test already covers verifying that we behave correctly for the HandleHolder emitted from codegen.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      handleHolder.readHandle
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "Handle readHandle has not been initialized in TestParticle yet."
+      )
+    }
+  }
+
+  @Test
+  fun handles_nameNotInEntitySpec_throwsNoElemException() {
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:45]: Similar for this test: it's only testing the structure that was created in this test.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    handleHolder.setHandle("writeHandle", secondHandle)
+
+    val actual = handleHolder.dispatcher
+
+    assertThat(actual).isSameInstanceAs(DUMMY_DISPATCHER_1)
+  }
+
+  @Test
+  fun getEntitySpec_handleNameNotInSpec_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:89]: Instantiating the handle holder this way is hiding details about the setup from the test body. While it's a bit more verbose, it would be preferable to simply do:
```
val handleHolder = HandleHolderBase(
  "dummyName",
  entitySpecs = mapOf("inSpec" to mock())
  )
  ```

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+    assertFailsWith<NoSuchElementException> {
+      handleHolder.getEntitySpecs("notInSpec")
+    }.also {
+      assertThat(it.message).isEqualTo(
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:94]: I think `assertFailsWith` has a variant that lets you provide the expected message as the first argument.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    val handleHolder = DummyHandleHolder()
+    val mock = mock<Handle>()
+
+    handleHolder.setHandle("writeHandle", mock)
+
+    assertThat(handleHolder.writeHandle).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap() {
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:166]: This test name feels a bit inaccurate, and could be broken into two tests:

`isEmpty_withNoHandlesSet_returnsFalse`

`isEmpty_withHandlesSet_returnsTrue`

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    handleHolder.readHandle.createForeignReference<EntityBase>(mock(), "someId")
+
+    verify(mock, times(1)).createForeignReference<EntityBase>(any(), any())
+  }
+
+  companion object {
+    val DUMMY_DISPATCHER_1 = Dispatchers.Default
+    val DUMMY_DISPATCHER_2 = Dispatchers.Main
+
+    class DummyHandleHolder : HandleHolderBase(
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:200]: A dummy constant is appropriate when the tests using it don't care at all about the details of the object that's created. For most of these tests, that's not the case. I'd recommend removing this completely in favor of constructing specific objects in each test, explicitly configuring them as described by that test.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+      "TestParticle",
+      mapOf(
+        "readHandle" to setOf(mock()),
+        "writeHandle" to setOf(mock()),
+        "entityButNotHandles" to setOf(mock())
+      )
+    ) {
+      val readHandle by handles
+      val writeHandle by handles
+      val handleButNotEntities by handles
```
Comment (jblebrun) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:210]: Related to the comment in the first two tests: this property-based structure is related to the generated sublcasses, not HandleHolderBase itself. So there's no need to create subclasses with properties like this, since we have test coverage for that elsewhere.

**Comment by jblebrun** [comment]
Looks like this is missing tests for `detach` and `reset`.

**Comment by jblebrun** [comment]
Some comments, but this is looking great so far!

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 2 above)
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:32]: You're right that this scenario is not likely to happen due to our generation, and it's covered by integration tests. However, isn't there value in defensive programming? And further, shouldn't we exercise our defenses, to ensure that the redundant safeguards are there? 

These checks are publicly accessible control flow that a user can exercise if they decided to hold the tool very wrong.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 1 above)
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:77]: Ah, I didn't realize that this is how dispatchers are actually used. I can change to test the interface and not the implementation.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 2 above)
Comment (jblebrun) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:32]: I think I didn't quite make my argument correctly here: it's that the only reason this test needs to exist is that you've created a structure inside of this test that behaves in a way that this test validates. 

There's nothing inherent to HandleHolderBase that implies that we should create properties and implement them by delegating to a map. That's an implementation detail of the code generation utility, and so we should include tests coincident with code actually generated by the utility.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 1 above)
Comment (jblebrun) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:77]: Well, your test is "technically correct" right now. I think the issue is really that we should fix the class to ensure that invariant...

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 6 above)
Comment (jblebrun) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:166]: (maybe inaccurate is not the right word... what I mean to say is that it's not really explaining the behavior that we care about, but rather, an internal implementation detail)

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 2 above)
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt]: Marking resolved: Removed these tests.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    handleHolder.setHandle("readHandle", mock())
+    handleHolder.setHandle("writeHandle", mock())
+    assertThat(handleHolder.handles).isNotEmpty()
+
+    handleHolder.reset()
```
Comment (mmandlis) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:247]: what happens / should happen if `reset` is called without `setHandle` being called before? do we have a test case that covers it (sorry, if i missed it)

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+    assertThat(handleHolder.handles).isEmpty()
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap_returnsTrue() {
```
Comment (mmandlis) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:253]: I'm not sure what `delegatesToHandlesMap` means, it think it is more confusing than helpful

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+  fun createForeignReference_someHandles_success() = runBlockingTest {
+    class DummyHandle : HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    ) {
+      val readHandle: Handle by handles
+    }
+
+    val handleHolder = DummyHandle()
+    val mock = mock<Handle>()
```
Comment (mmandlis) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:296]: maybe name this variable readHandleMock (to clearly distinguish between it and the entitySpec handle mock)?

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+    handleHolder.setHandle("readHandle", mock)
+
+    handleHolder.readHandle.createForeignReference<EntityBase>(mock(), "someId")
+
+    verify(mock, times(1)).createForeignReference<EntityBase>(any(), any())
```
Comment (mmandlis) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:301]: could you verify the mock call arguments as well?

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+      entitySpecs = mapOf("readHandle" to mock())
+    ) {
+      val readHandle: Handle by handles
+    }
+
+    val handleHolder = DummyHandle()
+    val mock = mock<Handle>()
+    handleHolder.setHandle("readHandle", mock)
+
+    handleHolder.readHandle.createForeignReference<EntityBase>(mock(), "someId")
```
Comment (mmandlis) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:299]: - sorry, i'm not sure i understand - why do you need the special DummyHandle class, instead of initializing the HandleHolderBase like in other tests?
- doesn't calling `handleHolder.readHandle.createForeignReference` actually tests the handle's createForeignReference call? why not `handleHolder.createForeignReference`?

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@OptIn(ExperimentalCoroutinesApi::class)
+@RunWith(JUnit4::class)
+class HandleHolderBaseTest {
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:29]: Jibbl and I have been discussing this in this thread. https://github.com/PolymerLabs/arcs/pull/6787#discussion_r545485464

What do you think?

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+    assertFailsWith(
+      IllegalArgumentException::class,
+      "TestParticle.writeHandle has already been initialized."
+    ) {
+      handleHolder.setHandle("writeHandle", mock())
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameNotInSpec_throwsNoElemException() {
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:143]: Thanks!

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.detach()
+
+    verify(readMock, times(1)).unregisterForStorageEvents()
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:190]: Thanks! That's good to know.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 9 above)
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:247]: That's a good test case, I'll write it up now. 

The idea I have for the test: before and after calling `reset`, the `handles` map should be empty.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 10 above)
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:253]: yeah, those don't make too much sense.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt (hunk 13 above)
Comment (alxrsngrtn) [reply javatests/arcs/sdk/HandleHolderBaseTest.kt:299]: Yes, thank you. The former was needed to do the latter, which as you pointed out, isn't correct.

File: javatests/arcs/sdk/HandleHolderBaseTest.kt [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock)
+    )
+
+    assertThat(handleHolder.getEntitySpecs("readHandle")).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun setHandle_whenAlreadyInitialized_throwsNoElemException() {
```
Comment (mmandlis) [inline javatests/arcs/sdk/HandleHolderBaseTest.kt:97]: nit - it throws IllegalArgumentException, but i think for test name it is sufficient to just state `_throwsException`

**Review by mmandlis — APPROVED** [review]
